lib/fauxpaas/open3_capture.rb

```diff
@@ -7,7 +7,11 @@ module Fauxpaas
   # Runner that uses Open3.capture3
   class Open3Capture
     def run(string)
-      Open3.capture3(string)
+      if string.include?("syslog")
```

Comment:
I think the fundamental question here is: whose responsibility is it to determine what kind of output we need from the system runner? It seems to me like for the syslog commands, it's the responsibility of `Fauxpaas::Cap` - for THESE kinds of commands, we need one kind of output, for THAT command, we need something else.

However, the system runner isn't exposed directly to `Fauxpaas::Cap` but rather through the `Fauxpaas::CapRunner` intermediate... What if `Fauxpaas::Cap` had subclasses that were responsible for knowing which backend runner they needed?

So I think this might look something like:

```ruby

class Instance
  ...

  # Maybe this could use Forwardable, not sure given the default
  # parameter, but maybe we could get rid of that?

  def deployer(fs = Filesystem.new)
    deploy_config(fs).deployer
  end

  def log_viewer(fs = Filesystem.new)
    deploy_config(fs).log_viewer
  end
end

class DeployConfig
  ...

  def deployer
    runner(Deployer)
  end

  def log_viewer
    runner(LogViewer)
  end

  private
  
  def runner(klass)
    klass.new(
      to_hash.merge("deployer_env" => Fauxpaas.deployer_env_root + deployer_env),
      appname
    )
  end
end

class Cap
  def initialize(options, stage, runner)
    ...
  end
end


class Deployer < Cap

  def initialize(options, stage, runner = CapRunner.new(Fauxpaas.system\_runner))
    super(options, stage, runner)
  end

  # deploy, rollback, restart, caches?
end

class LogViewer < Cap

  def initialize(options, stage, runner = CapRunner.new(KernelSystem.new))
    ...
  end

  # syslog\_view, follow, grep

end
```

and then the subcommands in `CLI::Syslog` would call `instance.log_viewer`, the subcommands in `CLI::Main` would call `instance.deployer`, and `Release#deploy` would call `deploy_config.deployer`

Note - I think `Deployer` is a better name than `Interrogator` with this split in mind since `Interrogator` implies something read-only. `Deployer` isn't a perfect name though - `deploy` and `rollback` are definitely related in that they are choosing which deployment to use and `caches` is pretty related to `rollback` and so makes sense here toor. `restart` is a little different though. I'm not sure it fits 100% with the `Deployer` name but I don't have anything better to use that encompasses those four actions.

This is a bit on the disruptive side though - if we think it's _too_ disruptive and we think we won't even be providing this functionality in the long term (b/c of centralized logging), I'm OK with living with the conditionality in `Open3Capture` for now or at least until we have other use cases for the (former) `KernelSystem`

Reply:
I was able to remove the conditional and push the job of choosing which one to use to the central service configuration. This is quite clean for main.rb, and what I'm going to call "acceptably dirty" in syslog.rb.  The larger issue you've uncovered is worth thinking about but I think we can punt on it for now.